src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Xml;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Definition;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Logging;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Run;
+
+namespace Microsoft.DotNet.Tools;
+
+/// <summary>
+/// Used to build a virtual project file in memory to support <c>dotnet run file.cs</c>.
+/// </summary>
+internal sealed class VirtualProjectBuildingCommand
+{
+    public Dictionary<string, string> GlobalProperties { get; } = new(StringComparer.OrdinalIgnoreCase);
+    public required string EntryPointFileFullPath { get; init; }
+
+    public int Execute(string[] binaryLoggerArgs, LoggerVerbosity verbosity)
+    {
+        var binaryLogger = GetBinaryLogger(binaryLoggerArgs);
+        var consoleLogger = new ConsoleLogger(verbosity);
```

Comment:
nit: should use new MSBuild API from @MichalPavlik to auto-detect the use of the Console Logger vs the Terminal Logger.

Reply:
Can you elaborate? I couldn't find such API.

Reply:
https://github.com/dotnet/msbuild/pull/11318

Reply:
Looks like the SDK intentionally doesn't use terminal logger now? I can call the same method so when that's fixed, the new logic will use the same logger:

https://github.com/dotnet/sdk/blob/7b44ce4a89e829b1b7ce2d1d8309289ad977da3d/src/Cli/dotnet/commands/dotnet-run/RunCommand.cs#L507-L508